### PR TITLE
KEYCLOAK-6630 Client scopes initial support

### PR DESCRIPTION
**Pull request by mposolda**

### CODE REVIEW

**Comment by mposolda** [comment]
I've added some tests to PR. It's ready for review now.

**Review by vramik — CHANGES_REQUESTED** [review]

**Comment by vramik** [comment]
in client-scope-mappings.html there is `<kc-tooltip>Client template has full scope allowed, which means this client will have the full scope of all roles.</kc-tooltip>` I'm not sure if it is actually used somewhere in admin console (I wasn't able to find it yet)

**Comment by vramik** [comment]
on Oracle I get: Error : 1408, Position : 70, Sql = CREATE INDEX DBALLO01.IDX_CLSCOPE_PROTMAP ON DBALLO01.PROTOCOL_MAPPER(CLIENT_SCOPE_ID), OriginalSql = CREATE INDEX DBALLO01.IDX_CLSCOPE_PROTMAP ON DBALLO01.PROTOCOL_MAPPER(CLIENT_SCOPE_ID), Error Msg = ORA-01408: such column list already indexed

**Comment by vramik** [comment]
on MSSQL I get ERROR [org.keycloak.connections.jpa.updater.liquibase.conn.DefaultLiquibaseConnectionProvider] (ServerService Thread Pool -- 53) Change Set META-INF/jpa-changelog-4.0.0.xml::4.0.0-KEYCLOAK-5579::mposolda@redhat.com failed.  Error: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'. [Failed SQL: ALTER TABLE [dbo].[CLIENT_TEMPLATE] DROP COLUMN [FULL_SCOPE_ALLOWED]]: liquibase.exception.DatabaseException: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'. [Failed SQL: ALTER TABLE [dbo].[CLIENT_TEMPLATE] DROP COLUMN [FULL_SCOPE_ALLOWED]]
...
...
Caused by: com.microsoft.sqlserver.jdbc.SQLServerException: The object 'DF_CLIENT_TEMPLATE_FULL_SCOPE_ALLOWED' is dependent on column 'FULL_SCOPE_ALLOWED'.

**Comment by mposolda** [comment]
@vramik Thanks for the review and DB testing! Agree with everything you pointed. Regarding client-scope-mappings.html, it looks that whole form "allowScope" (line 12-32 of file client-scope-mappings.html) can be removed. This page is shown when you are on detail of some client-scope in admin console and you click on tab "Scope" (the one with Role Scope Mappings).

**Comment by vramik** [comment]
@mposolda Regarding to Oracle, there is the [index](https://github.com/keycloak/keycloak/blob/fe98c30077d49822624c728e9fe668f8fb0ff074/model/jpa/src/main/resources/META-INF/jpa-changelog-3.2.0.xml#L152-L154) what it complains about

**Comment by mposolda** [comment]
Failing travis cross-dc test is due the https://issues.jboss.org/browse/KEYCLOAK-7017 . It's not related to this PR.

**Review by vramik — DISMISSED** [review]

**Comment by mposolda** [comment]
Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/389

**Comment by mposolda** [comment]
@fabianschlier yes, you're right. The client scopes reworks this topic and it will allow to limit used protocolMappers based on the scope parameter. Not sure when this will be available in keycloak master. Hopefully very soon (in the next release).

**Comment by mposolda** [comment]
@vramik I've did the rebase with latest master and needed to squash to single commit. Could you please re-run the DB matrix? You're welcome to re-review the PR again after rebase if you have time. But there are not much changes, mostly just the things related to the rebase.

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java [hunk 1]
```diff
@@ -1287,13 +1287,13 @@ public boolean removeGroup(GroupModel group) {
     }
 
     @Override
-    public List<ClientTemplateModel> getClientTemplates() {
-        if (isUpdated()) return updated.getClientTemplates();
-        List<String> clientTemplates = cached.getClientTemplates();
+    public List<ClientScopeModel> getClientScopes() {
+        if (isUpdated()) return updated.getClientScopes();
+        List<String> clientTemplates = cached.getClientScopes();
```
Comment (vramik) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java]: clientScopes

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java [hunk 2]
```diff
@@ -1287,13 +1287,13 @@ public boolean removeGroup(GroupModel group) {
...
         if (clientTemplates.isEmpty()) return Collections.EMPTY_LIST;
-        List<ClientTemplateModel> apps = new LinkedList<ClientTemplateModel>();
+        List<ClientScopeModel> apps = new LinkedList<ClientScopeModel>();
         for (String id : clientTemplates) {
-            ClientTemplateModel model = cacheSession.getClientTemplateById(id, this);
+            ClientScopeModel model = cacheSession.getClientScopeById(id, this);
             if (model == null) {
                 throw new IllegalStateException("Cached clientemplate not found: " + id);
```
Comment (vramik) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmAdapter.java]: clientScopes

File: model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/stream/HasRolePredicate.java [hunk 3]
```diff
@@ -55,8 +55,8 @@ public boolean test(Map.Entry<String, Revisioned> entry) {
             if (cachedClient.getScope().contains(role)) return true;
 
         }
-        if (value instanceof CachedClientTemplate) {
-            CachedClientTemplate cachedClientTemplate = (CachedClientTemplate)value;
+        if (value instanceof CachedClientScope) {
+            CachedClientScope cachedClientTemplate = (CachedClientScope)value;
```
Comment (vramik) [inline model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/stream/HasRolePredicate.java]: cachedClientTemplate

File: server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo2_3_0.java [hunk 4]
```diff
@@ -44,7 +44,7 @@ protected void migrateRealm(RealmModel realm) {
             MigrationUtils.updateProtocolMappers(client);
         }
 
-        for (ClientTemplateModel clientTemplate : realm.getClientTemplates()) {
+        for (ClientScopeModel clientTemplate : realm.getClientScopes()) {
```
Comment (vramik) [inline server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo2_3_0.java]: clientScope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java [hunk 5]
```diff
@@ -66,13 +66,13 @@
 
     void requireView(ClientModel client);
 
-    boolean canManage(ClientTemplateModel template);
+    boolean canManage(ClientScopeModel template);
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java [hunk 6]
```diff
@@ -66,13 +66,13 @@
...
 
-    void requireManage(ClientTemplateModel template);
+    void requireManage(ClientScopeModel template);
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java [hunk 7]
```diff
@@ -66,13 +66,13 @@
...
 
-    boolean canView(ClientTemplateModel template);
+    boolean canView(ClientScopeModel template);
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java [hunk 8]
```diff
@@ -66,13 +66,13 @@
...
 
-    void requireView(ClientTemplateModel template);
+    void requireView(ClientScopeModel template);
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissionEvaluator.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java [hunk 9]
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
     }
 
     @Override
-    public boolean canManage(ClientTemplateModel template) {
+    public boolean canManage(ClientScopeModel template) {
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java [hunk 10]
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
...
         return canManageClientsDefault();
     }
 
     @Override
-    public void requireManage(ClientTemplateModel template) {
+    public void requireManage(ClientScopeModel template) {
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java [hunk 11]
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
...
         if (!canManage(template)) {
             throw new ForbiddenException();
         }
     }
 
     @Override
-    public boolean canView(ClientTemplateModel template) {
+    public boolean canView(ClientScopeModel template) {
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java]: scope

File: services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java [hunk 12]
```diff
@@ -486,24 +486,24 @@ public void requireViewTemplates() {
...
         return canViewClientDefault();
     }
 
     @Override
-    public void requireView(ClientTemplateModel template) {
+    public void requireView(ClientScopeModel template) {
```
Comment (vramik) [inline services/src/main/java/org/keycloak/services/resources/admin/permissions/ClientPermissions.java]: scope

File: testsuite/utils/src/main/java/org/keycloak/testsuite/util/cli/TestCacheUtils.java [hunk 13]
```diff
@@ -48,8 +48,8 @@ public static void cacheRealmWithEverything(KeycloakSession session, String real
             cacheGroupRecursive(realm, group);
         }
 
-        for (ClientTemplateModel clientTemplate : realm.getClientTemplates()) {
-            realm.getClientTemplateById(clientTemplate.getId());
+        for (ClientScopeModel clientTemplate : realm.getClientScopes()) {
```
Comment (vramik) [inline testsuite/utils/src/main/java/org/keycloak/testsuite/util/cli/TestCacheUtils.java]: clientScope

**Comment by vramik** [comment]
@mposolda DB matrix passed

**Comment by mposolda** [comment]
@vramik Thanks, I've updated based on your feedback. I've changed all the methods on ClientPermissionEvaluator interface to reflect "clientScopes" instead of "templates" . Keep the changes in separate commit for easier review.

**Review by vramik — APPROVED** [review]

**Review by patriot1burke — APPROVED** [review]

**Comment by frekele** [comment]
Hi @mposolda ,

after upgrading keycloak 3.4.3.Final to v4.0.0-Final and soon after v4.1.0-Final, it is not possible to use Keycloak admin cli as in the example:
https://www.keycloak.org/docs/latest/server_development/index.html#example-using-java

```java
Keycloak keycloak = Keycloak.getInstance(
    "http://localhost:8080/auth",
    "master",
    "admin",
    "password",
    "admin-cli");
RealmRepresentation realm = keycloak.realm("master").toRepresentation();
```

Cause a NullPointerException in [BearerAuthFilter.java line 53](https://github.com/keycloak/keycloak/blob/4.1.0.Final/integration/admin-client/src/main/java/org/keycloak/admin/client/resource/BearerAuthFilter.java#L53)
```java
Caused by: java.lang.NullPointerException
	at org.keycloak.admin.client.resource.BearerAuthFilter.filter(BearerAuthFilter.java:53)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.filterRequest(ClientInvocation.java:573)
	... 45 more
```
authHeader is null because tokenService.grantToken call return all fields as empty in [TokenManager.java#L89](https://github.com/keycloak/keycloak/blob/4.1.0.Final/integration/admin-client/src/main/java/org/keycloak/admin/client/token/TokenManager.java#L89)

currenttoken

In keycloak 3.4.3.Final it works perfectly.


I do not know if this PR solve this problem, so I'm adding here my comment, did not want to open an issue because maybe this PR already solves this?!?!

Thank you

**Pull request merged**